### PR TITLE
fix: fix print message id without batch index

### PR DESCRIPTION
**Pull request by Chenxulin97**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #<xyz>

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation

*I found that different payload had the same message id*

### Modifications

*append batch index when call string method of MessageID*

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API: (**yes** / no)
  - The schema: (**yes** / no / don't know)
  - The default values of configurations: (**yes** / no)
  - The wire protocol: (**yes** / no)

### Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / GoDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar/impl_message.go [hunk 1]
```diff
@@ -184,7 +184,7 @@ func (id *messageID) BatchSize() int32 {
 }
 
 func (id *messageID) String() string {
-	return fmt.Sprintf("%d:%d:%d", id.ledgerID, id.entryID, id.partitionIdx)
+	return fmt.Sprintf("%d:%d:%d:%d", id.ledgerID, id.entryID, id.partitionIdx, id.batchIdx)
```
Comment (shoothzj) [inline pulsar/impl_message.go:187]: if batchIndex = -1, do we still need to print it?

Comment (RobertIndie) [reply pulsar/impl_message.go:187]: Let's keep it consistent with the Java client. Don't print the batch index when it's -1.

**Review by RobertIndie — COMMENTED** [review]
Please add the unit test case for it.

File: pulsar/impl_message.go (hunk 1 above)
Comment (Chenxulin97) [reply pulsar/impl_message.go:187]: ok

**Comment by shoothzj** [comment]
@RobertIndie, do we really need a unit test case for the `String()` method? In my opinion, simply testing the `String()` method has less value.

**Comment by RobertIndie** [comment]
> do we really need a unit test case for the `String()` method? In my opinion, simply testing the `String()` method has less value.

I think it's worth adding these tests. This is a public method exposed to users, and we need to ensure the correctness of the returned content. Moreover, other tests also rely on this String's output. We can see that this change has caused failures in other tests in the CI.

**Comment by shoothzj** [comment]
> > do we really need a unit test case for the `String()` method? In my opinion, simply testing the `String()` method has less value.
> 
> I think it's worth adding these tests. This is a public method exposed to users, and we need to ensure the correctness of the returned content. Moreover, other tests also rely on this String's output. We can see that this change has caused failures in other tests in the CI.

I understand the importance of writing test cases, and I won't insist on not writing them. However, I'd like to note that this part of the code logic is quite clear and straightforward.

**Comment by RobertIndie** [comment]
> However, I'd like to note that this part of the code logic is quite clear and straightforward.

We write tests not because the code is simple or complex, but to ensure the functionality behaves correctly. Even for simple code, I believe it's best to write tests to cover it. This is exactly why the metric of test coverage is important.

**Comment by shoothzj** [comment]
@RobertIndie Fair enough, thanks for your explanation. I won't block this review.

**Comment by geniusjoe** [comment]
I think the unit test case ran fail is because `TestGetMessagesByID` function in `pulsaradmin/pkg/admin/subscription_test.go` use message `id.String()` as a map key to check if current messages are in a same batch. 
When we change `id.String()` from `fmt.Sprintf("%d:%d:%d", id.ledgerID, id.entryID, id.partitionIdx)`
	to `fmt.Sprintf("%d:%d:%d:%d", id.ledgerID, id.entryID, id.partitionIdx, id.batchIdx)`, map key changed, so that the test case cannot distinguish messages in the same batch.

```
	for i := 0; i <= numberMessages; i++ {
		producer.SendAsync(ctx, &pulsar.ProducerMessage{
			Payload: []byte(fmt.Sprintf("hello-%d", i)),
		}, func(id pulsar.MessageID, message *pulsar.ProducerMessage, err error) {
			assert.Nil(t, err)
			messageIDMap[id.String()]++
			wg.Done()
		})
	}
```

If we need to fix test case, we can replace `messageIDMap[id.String()]` to `messageIDMap[fmt.Sprintf("%d:%d:%d", id.ledgerID, id.entryID, id.partitionIdx)]`, but I think this pr is an incompatibility change, so it should be greater to add change log in documentation.

**Comment by Chenxulin97** [comment]
It seems more works than I expect, I have no time to work on this.

**Pull request closed without merging**